### PR TITLE
Adds RubricAiEvaluation model

### PR DESCRIPTION
**Pull request by wilkie**
This follows the PR at #54360 which adds the migration to create the rubric_ai_evaluations table.

With this, the associated factories and such.

Our follow-up work will tie the rubric ai evaluations to the old learning_goal_ai_evaluations table.

## Links

- jira ticket: [AITT-272](https://codedotorg.atlassian.net/browse/AITT-272)

## Follow-up work

Task list for adding the `RubricAiEvaluation` model:

- [x] Add the migration (https://github.com/code-dot-org/code-dot-org/pull/54360)
- [x] Add the model (**THIS PR**) and start writing to the new table and the old
- [ ] Forge new `LearningGoalAiEvaluation` record (and move old one to new name) with a foreign key to point to `RubricAiEvaluation` (https://github.com/code-dot-org/code-dot-org/pull/54386)
- [ ] Add a data migration to move the old data from `LearningGoalAiEvaluation` records to this one (https://github.com/code-dot-org/code-dot-org/pull/54413)
- [ ] Drop old LearningGoalAiEvaluation table and properly add in logic to write/query RubricAiEvaluation (https://github.com/code-dot-org/code-dot-org/pull/54435)

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

**Comment by wilkie** [comment]
I'm not sure why these tests are timing out on the MySQL connection... there's no real reason.

**Comment by wilkie** [comment]
Again!! 😠

**Comment by wilkie** [comment]
Again.

```
ERROR["test_user_app_options_should_create_new_channel_if_one_doesn't_exist", "ApiControllerTest", 2529.548095286]
--
2573 | test_user_app_options_should_create_new_channel_if_one_doesn't_exist#ApiControllerTest (2529.55s)
2574 | Minitest::UnexpectedError:         Sequel::DatabaseError: Mysql2::Error::TimeoutError: Timeout waiting for a response from the last query. (waited 30 seconds)
2575 | app/helpers/levels_helper.rb:98:in `get_channel_for'
2576 | app/controllers/api_controller.rb:509:in `user_app_options'
2577 | lib/services/database_connections.rb:28:in `block in connect_to_reader'
2578 | app/helpers/read_replica_helper.rb:21:in `connected_to'
2579 | lib/services/database_connections.rb:27:in `connect_to_reader'
2580 | app/controllers/application_controller.rb:181:in `block in with_locale'
2581 | app/controllers/application_controller.rb:180:in `with_locale'
2582 | test/controllers/api_controller_test.rb:1019:in `block in <class:ApiControllerTest>'
2583 | test/testing/setup_all_and_teardown_all.rb:36:in `run'
```

**Comment by wilkie** [comment]
The `get_channel_for` times out predictably almost every test run. It boils down to this query: https://github.com/code-dot-org/code-dot-org/blob/b43061474de78b2641495fa232938a7115e6ae58/shared/middleware/helpers/storage_id.rb#L170-L172

I haven't touched any of that. I do have it referencing the project id, but not as a foreign key. There's no reason for this to happen.

Perhaps the cached test database is in a strange state where this lookup is very inefficient. It is hard to inspect.

It also has problems performing a `Project.create`, especially here: https://github.com/code-dot-org/code-dot-org/blob/b43061474de78b2641495fa232938a7115e6ae58/dashboard/app/models/backpack.rb#L27-L29 which the first line here (which seems to succeed, actually) also calls the `get_user_storage_id` function. It then times out on the `Project.create` on line 28. Baffling.

I have seen some reports of this behavior on our Slack. There is some knowledge that the tests and code related to Projects can cause problems completing test runs. Nobody has determined anything specific nor have a solution, it seems.

Locally testing the tests yields successful runs. I dropped and recreated the dashboard (and pegasus for good measure) databases. And ran just the one controller test at a time via `spring testunit`. Not exactly apples to apples.

**Comment by wilkie** [comment]
A couple of local test runs, succeeding:

```
cdodev@web:/app/src/dashboard$ bundle exec spring testunit ./test/controllers/learning_goal_teacher_evaluations_controller_test.rb
Running via Spring preloader in process 131290
Started with run options --seed 20019

  18/18: [==============================================================================================================================================================================================] 100% Time: 00:01:18, Time: 00:01:18

Finished in 110.76252s
18 tests, 60 assertions, 0 failures, 0 errors, 0 skips
cdodev@web:/app/src/dashboard$ bundle exec spring testunit ./test/controllers/learning_goal_teacher_evaluations_controller_test.rb
Running via Spring preloader in process 132611
Started with run options --seed 46788

  18/18: [==============================================================================================================================================================================================] 100% Time: 00:01:44, Time: 00:01:44

Finished in 110.72965s
18 tests, 60 assertions, 0 failures, 0 errors, 0 skips
```

**Comment by wilkie** [comment]
It does seem like adding `belongs_to :project` is incredibly problematic! The validations for the `project_id` field cause the timeouts on the software side, even though there are no such validations in the database specification. We could either remove the "belongs_to" or tell it to skip the validations.

You can skip the validations using `:optional`, which feels like the right thing to do.

```
belongs_to, optional: true
```

**Comment by wilkie** [comment]
I think I am going to cheat and just merge the next PR in the chain if this one is problematic with the test runs. It shouldn't hurt all that much to do two migrations... hopefully.

**Comment by wilkie** [comment]
Oh my goodness that worked!

File: dashboard/test/controllers/rubrics_controller_test.rb [hunk 1]
```diff
@@ -105,6 +105,8 @@ class RubricsControllerTest < ActionController::TestCase
     rubric = create :rubric
     learning_goal1 = create :learning_goal, rubric: rubric
     learning_goal2 = create :learning_goal, rubric: rubric
+    # TODO: save this to a variable and use it to create learning goal ai evaluations
+    create :rubric_ai_evaluation, user: student, requester: teacher, status: 1
```
Comment (davidsbailey) [inline dashboard/test/controllers/rubrics_controller_test.rb]: what breaks if you don't add this?

**Comment by davidsbailey** [comment]
I haven't dug up the slack threads, but I have personally run into problems with `stub_get_storage_id` though I don't recall the specifics. because ApiControllerTest is one of the two tests that uses it, I'm wondering if this is related to the failures you are seeing.

File: dashboard/test/controllers/rubrics_controller_test.rb (hunk 1 above)
Comment (wilkie) [reply dashboard/test/controllers/rubrics_controller_test.rb]: nothing should break if that is not there. it is there simply because the model was added and part of that process for this PR is to start writing to it.

**Comment by davidsbailey** [comment]
ok, now i am evaluating the possibility that we need to wrap any tests which use the projects factory like this: https://github.com/code-dot-org/code-dot-org/blob/188f29b6a16a0c700e8c110318d33b1b0f6874c3/dashboard/test/helpers/delete_accounts_helper_test.rb#L28

because rails test stuff makes sure to rollback any changes made through active record, but project owners are created using `DASHBOARD_DB` (Sequel) which are not automatically rolled back.

**Comment by wilkie** [comment]
> I haven't dug up the slack threads, but I have personally run into problems with stub_get_storage_id though I don't recall the specifics. because ApiControllerTest is one of the two tests that uses it, I'm wondering if this is related to the failures you are seeing.

Yes, stubbing that should work except it also would fail (and require stubbing out) `create_storage_id_for_user`

The thing is to remove all associations (including those in Rails, not just in the database) to the `Project` table. For some reason that is difficult to immediately understand, there will be a database timeout when creating a new `Project` id or querying it and performing the validations. It's all bizarre.

Avoiding it in my case was two-fold: I had to remove the association all the way down to flattening it to just `project_id` and also ensuring that my Factory for it didn't even create from the Project factory! The `belongs_to :project, optional: true` solved it in one case... because it avoided a lookup to validate the project id, but it still would fail when creating projects in the test factories... (but it was hard to tell what was going on, so it took me a little while)

This also failed with the same timeout locally! But only when I ran the whole test suite... it would pass individual test runs. Perplexing.

**Comment by wilkie** [comment]
> ok, now i am evaluating the possibility that we need to wrap any tests which use the projects factory like this:
> 
> https://github.com/code-dot-org/code-dot-org/blob/188f29b6a16a0c700e8c110318d33b1b0f6874c3/dashboard/test/helpers/delete_accounts_helper_test.rb#L28
> 
> because rails test stuff makes sure to rollback any changes made through active record, but project owners are created using `DASHBOARD_DB` (Sequel) which are not automatically rolled back.

I had also thought of something similar! The table for project_ids is not governed by Rails... and so it escapes, perhaps, some checks or the assumptions made aren't very sound. Since it is not wrapped the same way as other tables, it is also pretty hard to follow it when it goes belly up like this.

**Review by davidsbailey — APPROVED** [review]
looks like a good workaround on the project / mysql timeout issue. I'll see if I can test out the DASHBOARD_DB rollback idea to fix the original issue when I find a moment.
